009-ui-and-admin-server-security.md

```diff
@@ -0,0 +1,84 @@
+Like the UI flow, the credentials would be retrieved from the auth header and placed into the Kafka client properties as SCRAM credentials.
+
+## OAuth 2
+
+![Kafka Client using Oauth2](./images/oauth-client.png)
+
+In the OAuth 2 scenario, Strimzi currently requires the user to configure the Kafka client with a clientId and secret.
+The client will then obtain an access token from the identity provider which is KeyCloak and that token flows to the broker.
+In the broker, calls to KeyCloak will be made based on the type of token that was requested by the client and passed to the broker.
+
+![Strimzi UI using Oauth2](./images/oauth-ui.png)
+
+In the UI model, the user will be presented with a login dialogue box when the UI is loaded into the browser and requires the user to enter a userId and password.
+The userId needs to be pre-defined to KeyCloak through the KeyCloak user interface.
+When the user enters their userId and password, the UI will contact KeyCloak to authorise the user and obtain an access token.
+This process will be controlled by `passport.js`.
+KeyCloak will either return an access token or reject the request.
+If an access token is returned, it will be stored in the session data for the session.
+When the UI requires data from the Admin Server, the token is recovered from the session data and placed as an auth header on the request.
+The Admin Server receives the request and configures the Kafka client in preparation for calling the Kafka broker.
+The token is retrieved from the auth header and is added to the Kafka client JAAS configuration properties before executing the call to the Kafka broker.
+The Kafka broker, which is running a custom SASL plugin, retrieves the token and sets up the Kafka Principal for the user.
+The request then passes through the KeyCloak Authorizer for authorization.
+
+In order to call the Admin Server directly, the user would need to obtain a token from KeyCloak manually and then store that token in an auth header on the call to the Admin Server.
+The behaviour when the AdminServer received the request would be identical to the flow for the UI calling the Admin Server with a token.
+
+## Mutual TLS
+
+![Admin server using Mutual TLS](./images/mtls-admin.png)
+
+With Mutual TLS, authentication takes place in the TLS handshake between the external web client and the Admin Server.
+The server side can only detect a userId for the client so credentials cannot be propagated to the Kafka Broker that woud allow the broker to authenticate and authorize the request.
+This proposal recommends that the Kafka clients running under the Admin Server can use a valid Kafka User for the connection between the Kafka client and the broker and pass the external web client's userId to the broker login module.
+The Kafka broker then uses a SASL plugin to create a custom authentication module which sets up the Kafka principal which flows through to authorizer.
```

Comment:
I think this would be quite a lot of additional work and would affect also Kafka connections etc. So I wonder if any alternatives would be better solutions:
* Pass the cert/key from the UI in the login (via a form)
* Not support mutual TLS for the UI / API server at all

Reply:
So we're using some custom (EXTERNAL?) SASL mechansism and the broker-side doesn't really authenticate the identity asserted (it trusts it because of the TLS identity it's associated with), but builds a `KafkaPrincipal` using the asserted identity which is then used for authz?

Reply:
I removed the mTLS section altogether because it complicates the picture and, for the first implementation, I think it is sufficient to support no security, SCRAM-SHA and Oauth2. Mutual TLS can be revisited in the future if there is demand for it.